drivers/hwinfo/hwinfo_sam.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2019 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <hwinfo.h>
+#include <init.h>
+#include <soc.h>
+#include <string.h>
+
+static u8_t sam_uid[16];
+
+ssize_t _impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+{
+	if (length > sizeof(sam_uid)) {
+		length = sizeof(sam_uid);
+	}
+
+	memcpy(buffer, sam_uid, length);
+
+	return length;
+}
+
+/* On the Atmel SAM SoC series, the device id is located in the flash
+ * controller. The controller can either present the flash area containing
+ * the code, the unique identifier or the user signature area at the flash
+ * location. Therefore the function reading the device id must be executed
+ * from RAM with the interrupts disabled. To avoid executing this complex
+ * code each time the device id is requested, we do this at boot time at save
+ * the 128-bit value into RAM.
+ */
+__ramfunc static int hwinfo_sam_init(struct device *arg)
+{
+	Efc *efc = (Efc *)DT_FLASH_DEV_BASE_ADDRESS;
+	u8_t *flash = (u8_t *)CONFIG_FLASH_BASE_ADDRESS;
+	u32_t fmr;
+	int key;
+
+	/* Disable interrupts. */
+	key = irq_lock();
+
+	/* Disable code loop optimization and sequential code optimization. */
+	fmr = efc->EEFC_FMR;
+	efc->EEFC_FMR = (fmr & (~EEFC_FMR_CLOE)) | EEFC_FMR_SCOD;
+
+	/* Switch the flash controller to the unique identifier area. The flash
+	 * is not available anymore, hence we have to wait for it to be *NOT*
+	 * ready.
+	 */
+	efc->EEFC_FCR = EEFC_FCR_FKEY_PASSWD | EEFC_FCR_FCMD_STUI;
+	while ((efc->EEFC_FSR & EEFC_FSR_FRDY) == EEFC_FSR_FRDY) {
+		/* Wait */
+	}
+
+	/* Copy the 128-bit unique ID */
+	memcpy(sam_uid, flash, sizeof(sam_uid));
```

Comment:
Just to clarify: the branch to memcpy might switch to execute code from flash, no?

Reply:
You are correct. It works likely because the function is still in the cache, but it's not correct. I'll replace that with an explicit loop and a comment.